working/extension_structs/overview.md

```diff
@@ -0,0 +1,815 @@
+datatypes.  We expect that structs would be incorporated into switches with
+extended exhaustiveness checking as proposed in the patterns proposal in the
+obvious way.
+
+### Implementing structs
+
+It is an error to implement a struct.  Structs do not define interfaces.
+
+### Structs implementing interfaces
+
+Structs may implement interfaces, subject to the usual member conformance
+checks.
+
+### Semantics of structs
+
+Modulo generated members (see below) and identity (see below), structs behave
+semantically exactly as if they were de-sugared into classes in the obvious way.
+
+### Identity of structs
+
+The identity operator on structs *may* return `true` for structs such that:
+  - Both have the same runtime type
+  - For every pair of corresponding fields, identical could validly return true
+    on that pair.
+
+The identity operator on structs may always return false.
+
+In other words, the identity operator may be used as a "fast" cut-off for
+equality, but compilers are free to box and unbox structs at will without
+preserving identity.
+
+The intention is that the identity operator should simply serve as a fast check
+whether the two objects in question are "pointer equal", but it is valid for
+compilers to make other choices based on implementation concerns, pragmas, etc.
+
+
+### Generated members
+
+An abstract struct has no generated members.  For non-abstract structs, the
+following members are automatically derived by the compiler.
+
+#### Default constructor
+
+Every non-abstract struct defines a new private constructor with a hidden
+compiler generated name.  We refer to this constructor as the *generated primary
+constructor*.
+
+The generated primary constructor has a single positional parameter for every
+entry in the primary constructor list, each with the obvious type.
+
+If any entry has an initializer value provided, then every entry after (and
+including) that entry is an optional parameter with no default.
+
+For every entry with an initializer value, if no argument is passed for that
+parameter, the initializer value is assigned to that parameter in the
+initializer list of the constructor.  Note that this requires the ability to
+detect whether or not a parameter was passed, which is not expressible strictly
+as a de-sugaring.
+
+If no default constructor is defined in the class, a default constructor is
+generated which forwards to the generated primary constructor.
+
+**COMMENTARY(leafp):** *The treatment of initializers as default values here is
+  appealing, but it may be too cute.  It essentially adds non-const default
+  values only for this specific use case.  There's also a question of whether we
+  allow user defined constructors on the struct to also override the default
+  initializer, and if so via what syntax*.
+
+
+##### Alternative: named parameters
+
+These could be made named parameters.  This feels heavyweight, but has some
```

Comment:
It just makes the constructor call much more verbose.  Think `child` and `children` in Flutter: there is definitely documentation value to named parameters, but they also definitely add a lot of visual noise to the call site.

Reply:
For what it's worth, when I last dug into it, about half of constructors use positional parameters and half use named parameters. Thanks to the Flutter's framework design, named parameters have become very common in constructors.

(This is one of the main reason I've found it hard to come up with a nice data class syntax, because neither positional nor named parameters have a clear majority and supporting both is a hard syntax problem.)

Reply:
I'm not actually strongly married to either choice here.  The pitch of this proposal is we be opinionated and ship a single default, but this was one where I only weakly came down on this side.  And we could allow named with fairly minimal boilerplate, I think (just use braces in the primary constructor).

Reply:
Filed https://github.com/dart-lang/language/issues/2365 to discuss named vs positional.